pkg/gauges/contracts/BalancerTokenAdmin.sol

```diff
@@ -0,0 +1,364 @@
+        for (uint256 i = 0; i < numberOfSnapshotters; ++i) {
+            address snapshotter = _balancerToken.getRoleMember(snapshotRole, 0);
+            _balancerToken.revokeRole(snapshotRole, snapshotter);
+        }
+        // Give this contract snapshotting rights over the BAL token
+        _balancerToken.grantRole(snapshotRole, address(this));
+
+        // BalancerTokenAdmin now is the only holder of MINTER_ROLE and SNAPSHOT_ROLE for BalancerGovernanceToken.
+
+        // We can't prevent any other admins from granting other addresses these roles however.
+        // This undermines the ability for BalancerTokenAdmin to enforce the correct inflation schedule.
+        // The only way to prevent this is for BalancerTokenAdmin to be the only admin. We then remove all other admins.
+        uint256 numberOfAdmins = _balancerToken.getRoleMemberCount(adminRole);
+        uint256 skipSelf = 0;
+        for (uint256 i = 0; i < numberOfAdmins; ++i) {
+            address admin = _balancerToken.getRoleMember(adminRole, skipSelf);
+            if (admin != address(this)) {
+                _balancerToken.revokeRole(adminRole, admin);
+            } else {
+                // This contract is now the admin with index 0, we now delete the address with index 1 instead
+                skipSelf = 1;
+            }
+        }
+
+        // BalancerTokenAdmin doesn't actually need admin rights any more and won't grant rights to any more addresses
+        // We then renounce our admin role to ensure that another address won't gain absolute minting powers.
+        _balancerToken.revokeRole(adminRole, address(this));
+
+        // Perform sanity checks to make sure we're not leaving the roles in a broken state
+        require(_balancerToken.getRoleMemberCount(adminRole) == 0, "Address exists with admin rights");
+        require(_balancerToken.hasRole(minterRole, address(this)), "BalancerTokenAdmin is not a minter");
+        require(_balancerToken.hasRole(snapshotRole, address(this)), "BalancerTokenAdmin is not a snapshotter");
+        require(_balancerToken.getRoleMemberCount(minterRole) == 1, "Multiple minters exist");
+        require(_balancerToken.getRoleMemberCount(snapshotRole) == 1, "Multiple snapshotters exist");
+
+        // As BAL inflation is now enforced by this contract we can initialise the relevant variables.
+        startEpochSupply = _balancerToken.totalSupply();
+        startEpochTime = block.timestamp;
+        rate = _INITIAL_RATE;
+        emit MiningParametersUpdated(_INITIAL_RATE, startEpochSupply);
+    }
+
+    /**
+     * @notice Mint BAL tokens subject to the defined inflation schedule
+     * @dev Callable only by addresses defined in the Balancer Authorizer contract
+     */
+    function mint(address to, uint256 amount) external authenticate {
+        require(
+            _balancerToken.totalSupply().add(amount) <= _availableSupply(),
+            "Mint amount exceeds remaining available supply"
+        );
```

Comment:
CRV has an extra statement here:

```
    if block.timestamp >= self.start_epoch_time + RATE_REDUCTION_TIME:
        self._update_mining_parameters()
```

Reply:
The check is also quite cheap for minters that update before mint, right? As `start_epoch_time` will be warm.